src/app/games/trexo/TrexoRules.ts

```diff
@@ -0,0 +1,158 @@
+
+export class TrexoRules extends Rules<TrexoMove, TrexoState> {
+
+    private static instance: MGPOptional<TrexoRules> = MGPOptional.empty();
+
+    private static getOwner(piece: TrexoPieceStack): PlayerOrNone {
+        return piece.getOwner();
+    }
+    private static isInRange(coord: Coord): boolean {
+        return coord.isInRange(TrexoState.SIZE, TrexoState.SIZE);
+    }
+    public static readonly TREXO_HELPER: NInARowHelper<TrexoPieceStack> =
+        new NInARowHelper(TrexoRules.isInRange, TrexoRules.getOwner, 5);
+
+    public static get(): TrexoRules {
+        if (TrexoRules.instance.isAbsent()) {
+            TrexoRules.instance = MGPOptional.of(new TrexoRules());
+        }
+        return TrexoRules.instance.get();
+    }
+    private constructor() {
+        super(TrexoState);
+    }
+    public applyLegalMove(move: TrexoMove, state: TrexoState): TrexoState {
+        return state
+            .drop(move.getZero(), Player.ZERO)
+            .drop(move.getOne(), Player.ONE)
+            .incrementTurn();
+    }
+    public isLegal(move: TrexoMove, state: TrexoState): MGPFallible<void> {
+        if (this.isUnevenGround(move, state)) {
+            return MGPFallible.failure(TrexoFailure.CANNOT_DROP_PIECE_ON_UNEVEN_GROUNDS());
+        }
+        if (this.landsOnOnlyOnePiece(move, state)) {
+            return MGPFallible.failure(TrexoFailure.CANNOT_DROP_ON_ONLY_ONE_PIECE());
+        }
+        return MGPFallible.success(undefined);
+    }
+    public isUnevenGround(move: TrexoMove, state: TrexoState): boolean {
+        const zero: TrexoPieceStack = state.getPieceAt(move.getZero());
+        const one: TrexoPieceStack = state.getPieceAt(move.getOne());
+        return zero.getHeight() !== one.getHeight();
+    }
+    public landsOnOnlyOnePiece(move: TrexoMove, state: TrexoState): boolean {
+        const zeroSpace: TrexoPieceStack = state.getPieceAt(move.getZero());
+        const oneSpace: TrexoPieceStack = state.getPieceAt(move.getOne());
+        if (zeroSpace.getUpperTileId() === -1 && oneSpace.getUpperTileId() === -1) {
+            return false;
+        }
+        return zeroSpace.getUpperTileId() === oneSpace.getUpperTileId();
+    }
+    public static getSquareScore(state: TrexoState, coord: Coord): number {
+        return TrexoRules.TREXO_HELPER.getSquareScore(state, coord);
+    }
+    public static getVictoriousCoords(state: TrexoState): Coord[] {
+        const victoryOfLastPlayer: Coord[] = [];
+        const victoryOfNextPlayer: Coord[] = [];
+        const lastPlayer: Player = state.getCurrentOpponent();
+        for (let x: number = 0; x < TrexoState.SIZE; x++) {
+            // for every column, starting from the bottom of each column
+            for (let y: number = 0; y < TrexoState.SIZE; y++) {
+                // while we haven't reached the top or an empty space
+                const coord: Coord = new Coord(x, y);
```

Comment:
comme ça c'est dupliqué au moins trois fois (ici, juste en dessous, et dans le minimax), tu peux peut-être faire une méthode qui itère sur le plateau dans TrexoState et l'utiliser ici ?

Reply:
j'pourrais utiliser 
````
    [Symbol.iterator](): IterableIterator<P> {
        const linedUpElements: P[] = [];
        for (const lines of this.board) {
            linedUpElements.push(...lines);
        }
        return linedUpElements.values();
    }
````
Qui itère sur les valeurs mais dans certains cas il faut la coord en plus 8
une idée de comment avoir les deux (itère sur coords, itère sur valeurs)  ?
Ce serait à mettre dans GameStateWithTable naturellement !

Reply:
tu fais un itérateur de P, et un itérateur de [Coord, P] simplement ?
et un .iteratorWithCoords() qui retourne le bon itérable, s'tu vois c'que jveux dire ! (sinon ask me IRL)

Reply:
IRL, genre en mai :o ! 
Non voilà j'ai fait un toMap() qui itère sur {key: Coord, value: P}